### DISCUSSION TITLE
Why MAX_MAXZOOM = 15?

### DISCUSSION THREAD

**Discussion by tyrauber** [onthegomap/planetiler#506] [Q&A] [answered]
I've ran through a planet wide build, and in testing, everything works great up until zoom level 16.  At which the whole map goes grey.  I tried tweaking the `.setZoomRange(10, 24);` per layer on the build profile, but that raised the error: 

> Max zoom must be <= 15.

In looking through the code I noticed a hard coded limit to max zoom at 15, MAX_MAXZOOM.

https://github.com/onthegomap/planetiler/blob/1375ee7abb93a04df786d31b45cc2a34d3200841/planetiler-core/src/main/java/com/onthegomap/planetiler/config/PlanetilerConfig.java#L57

Is this a performance compromise? Or to limit build time or file size?

I made the change locally and was able to build a small region, no problem.

All the map libraries support up to a max zoom of 24. Why is Planetiler limiting the zoom to 16?

I found by setting the MAX_MAX_ZOOM to 23,  I didn't need to tweak the max zoom level for the layers, they properly scaled up.

Why MAX_MAXZOOM = 15?

**Comment by bdon** [comment, marked as answer]
Hi Ty,

You should enable overzooming in your map: the `source` object in the map style should be set to 15, like here: https://github.com/protomaps/basemaps/blob/main/index.html#L45

More generally, MAX_MAXZOOM=15 is a technical limitation of Planetiler's 32-bit internal tile addressing. However, zoom 15 is sufficient for detailed mapping of typical street-level datasets when combined with overzooming, anywhere from zoom 16-20 should look fine as the renderer is re-using the zoom 15 tile.

**Reply by tyrauber** [reply]
Hi Brandon, I am not seeming overzooming. Even with `maxZoom: 24`, the map still blanks out above 16.

I am *not* seeing this behavior with `protomaps_vector_planet_odbl_z10.pmtiles` though, just the `planet-daylight-all-v1.23` I produced, which suggests to me it's a build profile issue.  I have a `planet-230227.osm.pmtiles` building which I can test shortly, but I suspect the same issue there as well.

**Reply by tyrauber** [reply]
Sorry, comprehension failure.  Adding `maxZoom: 15` to the **source object** enables overzooming.  Thank you @bdon!

**Reply by msbarry** [reply]
Thanks @bdon! It would be possible to make planetiler go higher than z15 by spilling over and using more bits but z15 should be sufficient for most applications. The benefit would be to make the z15 tiles a bit smaller by omitting some features that your style only shows when zoomed into z16+, but the cost would be that your planet archive would weigh closer to 1TB+ and take longer to generate. For comparison, openmaptiles vector tiles go to z14, mapbox streets vector tiles go to z16 and apple maps vector tiles go to z17.

**Reply by tyrauber** [reply]
@msbarry, I am not sure it's necessary. The overzooming works great once enabled 😓.

**Reply by YohanSciubukgian** [reply]
@msbarry Could you reconsider moving to a higher MaxZoom level and stick to a default z15 that could be sufficient for majority of use cases?
Do you still have the 32 bits limitations as of today?
I'm fine with the tradeoffs on having a much higher output pmtiles file as far as it provides more fine-grained details on the map.

Ex: Having maxzoom level up to z18 provide well detailed street POI (such as stores/restaurants/etc.) entry locations which is harder to represent at z15 (or not possible).
- z15: https://www.maptiler.com/google-maps-coordinates-tile-bounds-projection/#15/2.35/48.86
- z18: https://www.maptiler.com/google-maps-coordinates-tile-bounds-projection/#18/2.35/48.86

I saw that you are also working on supporting OuvertureMaps, which looks to be a promising project.
As OuvertureMaps split data per dimensions (Places, Placenames, Buildings, Roads, Base (Land, Land Use & Water)), could we have a MaxZoom level configured per OuvertureMaps data dimension based on our use cases to have multiple smaller pmtile files where any of them could fit in the 32 bits limitation?

**Reply by bdon** [reply]
You should be relying on overzooming in the client/style for zooms above 15, not data tiles existing above 15.

**Reply by YohanSciubukgian** [reply]
@bdon Thanks for the quick answer.
I understand that it's already the solution suggested on this thread.

I'm not sure to understand where the limitation is as I'm new to maps and all concepts around its tooling.
It looks like OpenStreetMap provides Tiles up to ZoomLevel=20 and where ZoomLevel=17 looks to be the right level for managing addresses as per their [OpenStreetMap Wiki Zoom_levels](https://wiki.openstreetmap.org/wiki/Zoom_levels)
![image](https://github.com/onthegomap/planetiler/assets/30290851/741ebd21-d9d0-4f68-bbac-ec527b7736f5)

AFAIU, OuvertureMaps is based on OpenStreetMaps and they are adding even more details based on big tech providers to have better global coverage of the map, also adding 3D buildings and more. Don't they also plan to add more zoomLevel details on existing OpenStreetMap data if there is any ZoomLevel 15 limitation today?

I'm looking for building a map which could provide such level of details without sacrificing user XP with overzooming feature (I understand that overzooming will not add more details but just zoom on the existing vector rendered at level 15) and using a pmtile vector tile format for ease of hosting (if it could also fit this need).

As you might see I'm a little bit confuse about your answer, could you provide me some guidance please?

---
UPDATE:
I also check configuration of [felt/tippecanoe: Build vector tilesets from large collections of GeoJSON features. (github.com)](https://github.com/felt/tippecanoe?tab=readme-ov-file#zoom-levels) and it looks like it supports creatin vector tiles up to zoomLevel=22. What would be the need for providing support for generating vector Tiles for z=22 if there is not data for such level  of details? Is there any tradeoff/limitations that I didn't get from first pass on these oss projects?
![image](https://github.com/onthegomap/planetiler/assets/30290851/f92ff3c0-9032-42d3-8ce7-bca345d4acc8)

---
UPDATE 2:
I have tested to download the prebuild basemap pmtile from [Builds | Protomaps Basemaps](https://maps.protomaps.com/builds/) and I see that even if the maxZoomLevel is limited 15, the overzooming to level 17 seems to fit my need.
I have used the following [PMTiles Viewer (protomaps.github.io)](https://protomaps.github.io/PMTiles/) tool to test the overzooming visualization and take the following screenshot:
![image](https://github.com/onthegomap/planetiler/assets/30290851/1751088a-241a-48eb-843d-ab0881126b31)

**Reply by bdon** [reply]
Yes, the PMTiles Viewer as you have shown in the screenshot is displaying zoom level 17 using over zoomed 15 tiles. 15 is generally enough for POI and building-entrance level mapping.

**Reply by YohanSciubukgian** [reply]
Thanks for the confirmation.
I'll start with maxZoomLevel 15 with overzooming as you suggest it.

For my personal understanding, could you let me know what & where is the limitation from providing more detailed maxZoomLevel (> 15)? Do you have any link / github issue that I could learn more from?

**Reply by bdon** [reply]
See https://github.com/onthegomap/planetiler/pull/266 and related issues

**Comment by joejuzl** [comment]
How difficult would it be to enable zoom 16 tile generation in planetiler?
I'm in a situation where tile size (in bytes) is a big factor, and where we are always displaying at high zoom levels. Being able to generate tiles at zoom level 16 would be very helpful.

Is it just the case of changing a variable from `int` to `long`? Or are there wide reaching implications in the code-base?

**Reply by msbarry** [reply]
The main constraint is that all features get sorted by a 64-bit key that gets encoded like this: https://github.com/onthegomap/planetiler/blob/main/planetiler-core/src/main/java/com/onthegomap/planetiler/collection/FeatureGroup.java#L110. There are 5.7 billion z16 tiles, which we only need one extra bit (33 total) so we'd just need to steal a bit from one of the other components.

**Reply by erik** [reply]
Not sure if this is a common use, but both @joejuzl and [StrandedKitty in Slack](https://osmus.slack.com/archives/C031V9E9RMG/p1685131269178009) were interested in a fixed zoom level at z16. I think we can already support that particular case, since z16 would have `4**16 == 2**32` tiles, which can fit into the 32 bits if no other levels are needed.

We'd need to adjust the logic for checking the [min and max zoom](https://github.com/onthegomap/planetiler/blob/1375ee7abb93a04df786d31b45cc2a34d3200841/planetiler-core/src/main/java/com/onthegomap/planetiler/config/PlanetilerConfig.java#L67) bounds, but that might be the only necessary change?

**Reply by bdon** [reply]
IMHO the benefits of moving to true z0-z16 support outweigh the simplicity of ID-remapping workarounds that do only one zoom level.

[QA tiles](https://osmlab.github.io/osm-qa-tiles/) ( [planetiler example](https://github.com/onthegomap/planetiler/blob/main/planetiler-examples/src/main/java/com/onthegomap/planetiler/examples/OsmQaTiles.java) ) which are used by some analytics apps are the main single-zoom application I know of.

**Reply by erik** [reply]
On second thought, yeah I agree. Not worth doing the hacky solution here. I was thinking this could be a nice stopgap until true z16 support is implemented, but other changes are required for this anyway (like some kind of hacky change to [TileCoord](https://github.com/onthegomap/planetiler/blob/main/planetiler-core/src/main/java/com/onthegomap/planetiler/geo/TileCoord.java))

**Comment by adjourn** [comment]
To signal there's still interest few years later - 👋 

Data is only growing and zoom 15 is getting up there weight wise, with common POIs, buildings, street names, address numbers, etc basic stuff. Wouldn't mind the ability to offload some of the less important but still needed stuff to zoom 16 tiles.

**Reply by adjourn** [reply]
Lower data transfer is a bonus but..

Found one of my old phones, probably something like 5 year old Android and decided to give it a go - it struggled in dense cities, _especially_ zoom 15. Then there's the possibility of bad internet.. Would really love to improve on these two areas some day.

**Comment by msbarry** [comment]
#1523 adds support for generating tiles to z16

**Reply by coolusaHD** [reply]
@msbarry  I'm just curious as I'm also new to this topic. If you updated from int to long why set the limitation to z16 when with the larger adresses also z20 as the osm limit would also be possible? Or do I miss something ?

**Reply by msbarry** [reply]
Planetiler sorts all features in the planet by 64-bit long that contains tile ID, layer, and feature sort key within that tile:

https://github.com/onthegomap/planetiler/blob/2f434854f518438333978c10bcdb620a347d2f60/planetiler-core/src/main/java/com/onthegomap/planetiler/collection/FeatureGroup.java#L173

So when we added one bit to tile IDs, we had to take a bit away from feature sort key, which reduced their resolution a bit.  Also within each zoom level we use a few optimized 32-bit data utilities (roaring bitmaps, and hilbert index) which just barely fit all z16 tiles, so going past z16 would make things slower and less efficient for everyone.

z16 and higher also ends up mostly duplicating features from the level below it, which increases processing, storage, and bandwidth requirements and isn't necessary since clients can overzoom the highest zoom level in the tileset. It makes more sense for systems that store geometries once and generate tiles from them on the fly like baremaps and only helps if the profile actual filters enough features out z14/z15 tiles that re-downloading the duplicated features is worth it for clients.
